// rotate/rotate.js
#!/usr/bin/env node
/*jslint nodejs:true */

var
	http = require("http"),
	url = require("url"),
	argv = require("optimist").string("h", "s").argv;

var a = "a".charCodeAt(0);
var z = "z".charCodeAt(0);
var A = "A".charCodeAt(0);
var Z = "Z".charCodeAt(0);

function rotChar(c) {
	var
		plainbyte = c.charCodeAt(0),
		cipherbyte = plainbyte;

	if (plainbyte >= a && plainbyte <= z) {
		cipherbyte = (((plainbyte - a) + 13) % 26) + a;
	}
	else if (plainbyte >= A && plainbyte <= Z) {
		cipherbyte = (((plainbyte - A) + 13) % 26) + A;
	}

	return String.fromCharCode(cipherbyte);
}

exports.rotChar = rotChar;

function rot13(plaintext) {
	var ciphertext = "";

	for (var i in plaintext) {
		ciphertext += rotChar(plaintext[i]);
	}

	return ciphertext;
}

exports.rot13 = rot13;

function page(req) {
	var
		query = url.parse(req.url, true).query,
		text = query.text || "",
		ciphertext = rot13(text);

	return "<!DOCTYPE html>" +
		"<head>" +
		"<title>ROT13</title><meta http-equiv=\"Content-Type\" content=\"text/html;charset=utf-8\" />" +
		"</head>" +
		"<body onload=\"document.getElementById('text').focus();\">" +
		"<style type=\"text/css\" />" +
		"body { text-align: center; }" +
		"a { color: #000; }" +
		"a:visited { color: #000; }" +
		"#logo { font-size: 500%; }" +
		"#wc_form { display: block; margin: 0 auto; width: 400px; }" +
		"label { display: block; float: left; width: 5%; padding-right: 1%; }" +
		"#github { position: fixed; bottom: 1em; right: 1em; }" +
		"</style>" +
		"<h1 id=\"logo\"><a href=\"/\">ROT13</a></h1>" +
		"<div id=\"rot_form\">" +

		"<form action=\"\" method=\"get\">" +

		"<p><textarea id=\"text\" name=\"text\" cols=\"50\" rows=\"15\">" + ciphertext + "</textarea></p>" +

		"<p><button type=\"submit\">Rotate</button></p>" +

		"</div>" +
		"<div id=\"github\"><a href=\"https://github.com/mcandre/node-rotate\">GitHub</a></div>" +
		"</body></html";
}

function server() {
	http.createServer(function (req, res) {
		res.writeHead(200, {"Content-Type": "text/html"});
		res.end(page(req));
	}).listen(6125, "localhost");

	console.log("Server running at http://localhost:6125/");
}

exports.server = server;

function commandLine() {
	process.stdin.resume();
	process.stdin.setEncoding("UTF8");

	process.stdin.on("data", function(chunk) {
		process.stdout.write(rot13(chunk));
	});

	process.stdin.on("end", function() {});
}

exports.commandLine = commandLine;

function usage() {
	console.log("rotate.js [options]");
	console.log("-s\tServer");
	console.log("-h\tUsage");
}

exports.usage = usage;

function main() {
	if ("h" in argv) {
		usage();
	}
	else if ("s" in argv) {
		server();
	}
	else {
		commandLine();
	}
}

if (!module.parent) { main(); }
